engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java

```diff
@@ -1,6 +1,30 @@
 package org.enso.interpreter.runtime.data;
 
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.TruffleLanguage;
+import com.oracle.truffle.api.interop.InteropLibrary;
 import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.EnsoLanguage;
 
 /** All non-primitive Enso types extends from {@code EnsoObject}. */
-public interface EnsoObject extends TruffleObject {}
+@ExportLibrary(InteropLibrary.class)
+public abstract class EnsoObject implements TruffleObject {
+  @ExportMessage
+  public boolean hasLanguage() {
+    return true;
+  }
+
+  @ExportMessage
+  public Class<? extends TruffleLanguage<?>> getLanguage() {
+    return EnsoLanguage.class;
+  }
+
+  @ExportMessage
+  @TruffleBoundary
+  public Object toDisplayString(boolean allowSideEffects) {
```

Comment:
This method can be `abstract`. I suggest to make it `abstract`. Definitely don't delegate to `toString()` - that's just a way for the guest language hacker to gain too much information about Enso interpreter internals.

Reply:
Refactored to `abstract` in 46ebbd3e3fd5d1cf8e2e25be8076467d1d41a518 . Note that it must be `public`, otherwise Truffle DSL errors with something like:
```
[error] /home/pavel/dev/enso/engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/Warning.java:24:1:  error: Found invisible exported elements in super type 'EnsoObject': 
[error] public final class Warning extends EnsoObject {
[error]              ^     - org.enso.interpreter.runtime.data.EnsoObject.toDisplayString(boolean)
[error]   Increase their visibility to resolve this problem.
```